### PR TITLE
Refactor evaluators and add correctness evaluator

### PR DESCRIPTION
**Pull request by Disiok**
## Description
- Refactor existing evaluators to have consistent interface.
  - Note that this is a breaking change for anyone currently using evaluators.
- Add correctness evaluator (depends on labeled answer) 

## Additional unrelated fixes
- update prompt templates to run output parser formatting if output parser is set.

### CODE REVIEW

File: llama_index/evaluation/base.py
Comment (logan-markewich) [inline llama_index/evaluation/base.py:1]: Can we add async as well?

File: llama_index/program/llm_program.py [hunk 1]
```diff
@@ -49,14 +48,10 @@ def from_defaults(
...
         if prompt_template_str is not None:
             prompt = PromptTemplate(prompt_template_str)
-        function_call = function_call or {
-            "name": output_parser.output_cls.schema()["title"]
-        }
         return cls(
             output_parser,
             prompt=cast(PromptTemplate, prompt),
             llm=llm,
-            function_call=function_call,
```
Comment (logan-markewich) [inline llama_index/program/llm_program.py:59]: I'm guessing this was unused?

**Review by logan-markewich — APPROVED** [review]
This looks good to me. I hope the clear base class encourages more integrations.

My only comment is that it would be nice to have both async and some kind of batch evaluation.

Right now, each evaluator is for a single query/response. Some kind of `EvalRunner` abstraction that runs any number of evaluators in some kind of async pool would help a lot for usability.

**Comment by logan-markewich** [comment]
Just pinging @siddhant01 as a heads up on this change, as they seem to be using our eval stuff in production apps :)

File: llama_index/evaluation/base.py
Comment (Disiok) [reply llama_index/evaluation/base.py:1]: will do in future PR.

File: llama_index/program/llm_program.py (hunk 1 above)
Comment (Disiok) [reply llama_index/program/llm_program.py:59]: ya looks like it.

**Comment by Disiok** [comment]
> This looks good to me. I hope the clear base class encourages more integrations.
> 
> My only comment is that it would be nice to have both async and some kind of batch evaluation.
> 
> Right now, each evaluator is for a single query/response. Some kind of `EvalRunner` abstraction that runs any number of evaluators in some kind of async pool would help a lot for usability.

Great points, let's add in future PRs.

**Pull request merged**